src/platform_impl/windows/event_loop/runner.rs

```diff
@@ -51,55 +56,70 @@ pub(crate) enum RunnerState {
+    WakeUp,
 }
 
 impl EventLoopRunner {
-    pub(crate) fn new(thread_msg_target: HWND) -> EventLoopRunner {
-        EventLoopRunner {
+    pub(crate) fn new(thread_id: u32, thread_msg_target: HWND) -> Self {
+        Self {
+            thread_id,
             thread_msg_target,
             interrupt_msg_dispatch: Cell::new(false),
             runner_state: Cell::new(RunnerState::Uninitialized),
             control_flow: Cell::new(ControlFlow::default()),
             exit: Cell::new(None),
             panic_error: Cell::new(None),
             last_events_cleared: Cell::new(Instant::now()),
-            event_handler: Cell::new(None),
+            event_handler: Rc::new(Cell::new(None)),
             event_buffer: RefCell::new(VecDeque::new()),
         }
     }
 
-    /// Associate the application's event handler with the runner
+    /// Associate the application's event handler with the runner.
     ///
     /// # Safety
-    /// This is ignoring the lifetime of the application handler (which may not
-    /// outlive the EventLoopRunner) and can lead to undefined behaviour if
-    /// the handler is not cleared before the end of real lifetime.
     ///
-    /// All public APIs that take an event handler (`run`, `run_on_demand`,
-    /// `pump_events`) _must_ pair a call to `set_event_handler` with
-    /// a call to `clear_event_handler` before returning to avoid
-    /// undefined behaviour.
-    pub(crate) unsafe fn set_event_handler<F>(&self, f: F)
-    where
-        F: FnMut(Event),
-    {
-        // Erase closure lifetime.
-        // SAFETY: Caller upholds that the lifetime of the closure is upheld.
-        let f =
-            unsafe { mem::transmute::<Box<dyn FnMut(Event)>, Box<dyn FnMut(Event)>>(Box::new(f)) };
+    /// The returned type must not be leaked (as that would allow the application to be associated
+    /// with the runner for too long).
+    pub(crate) unsafe fn set_app<'app>(
+        &self,
+        app: &'app mut (dyn ApplicationHandler + 'app),
+    ) -> impl Drop + 'app {
+        // Erase app lifetime, to allow storing on the event loop runner.
+        //
+        // SAFETY: Caller upholds that the lifetime of the closure is upheld, by not dropping the
+        // return type which resets it.
+        let f = unsafe {
+            mem::transmute::<
+                &'app mut (dyn ApplicationHandler + 'app),
+                &'static mut (dyn ApplicationHandler + 'static),
+            >(app)
+        };
+
         let old_event_handler = self.event_handler.replace(Some(f));
+
         assert!(old_event_handler.is_none());
-    }
 
-    pub(crate) fn clear_event_handler(&self) {
-        self.event_handler.set(None);
+        struct Resetter(Rc<EventHandler>);
+
+        impl Drop for Resetter {
+            fn drop(&mut self) {
+                self.0.set(None);
+            }
+        }
+
+        Resetter(self.event_handler.clone())
     }
```

Comment:
Can we move it outside of the function? It's private anyway, so no need to test it?

Reply:
I'd prefer not, it's easier to ensure soundness of the above `unsafe` when you can see exactly _where_ we're resetting / clearing the event handler again.